File: backend-node/models/procurement.js
const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  const Supplier = sequelize.define('Supplier', {
    supplier_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    supplier_code: {
      type: DataTypes.STRING(50),
      unique: true,
      allowNull: false,
    },
    name: {
      type: DataTypes.STRING(200),
      allowNull: false,
    },
    contact_person: {
      type: DataTypes.STRING(100),
    },
    email: {
      type: DataTypes.STRING(100),
    },
    phone: {
      type: DataTypes.STRING(20),
    },
    address: {
      type: DataTypes.TEXT,
    },
    city: {
      type: DataTypes.STRING(100),
    },
    state: {
      type: DataTypes.STRING(100),
    },
    country: {
      type: DataTypes.STRING(100),
    },
    postal_code: {
      type: DataTypes.STRING(20),
    },
    tax_id: {
      type: DataTypes.STRING(50),
    },
    payment_terms: {
      type: DataTypes.STRING(100),
    },
    credit_limit: {
      type: DataTypes.DECIMAL(15, 2),
    },
    rating: {
      type: DataTypes.DECIMAL(3, 2),
    },
    status: {
      type: DataTypes.ENUM('active', 'inactive', 'blacklisted'),
      defaultValue: 'active',
    }
  }, {
    tableName: 'suppliers',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  const PurchaseOrder = sequelize.define('PurchaseOrder', {
    po_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    po_number: {
      type: DataTypes.STRING(50),
      unique: true,
      allowNull: false,
    },
    supplier_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    order_date: {
      type: DataTypes.DATEONLY,
      allowNull: false,
    },
    expected_delivery_date: {
      type: DataTypes.DATEONLY,
    },
    actual_delivery_date: {
      type: DataTypes.DATEONLY,
    },
    status: {
      type: DataTypes.ENUM('draft', 'pending_approval', 'approved', 'ordered', 'partially_received', 'received', 'cancelled'),
      defaultValue: 'draft',
    },
    total_amount: {
      type: DataTypes.DECIMAL(15, 2),
    },
    tax_amount: {
      type: DataTypes.DECIMAL(15, 2),
    },
    discount_amount: {
      type: DataTypes.DECIMAL(15, 2),
    },
    net_amount: {
      type: DataTypes.DECIMAL(15, 2),
    },
    payment_terms: {
      type: DataTypes.STRING(100),
    },
    notes: {
      type: DataTypes.TEXT,
    },
    approved_by: {
      type: DataTypes.INTEGER,
    },
    approved_at: {
      type: DataTypes.DATE,
    },
    created_by: {
      type: DataTypes.INTEGER,
      allowNull: false,
    }
  }, {
    tableName: 'purchase_orders',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  const POItem = sequelize.define('POItem', {
    item_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    po_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    material_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    quantity: {
      type: DataTypes.DECIMAL(15, 3),
      allowNull: false,
    },
    unit_price: {
      type: DataTypes.DECIMAL(15, 2),
      allowNull: false,
    },
    total_price: {
      type: DataTypes.DECIMAL(15, 2),
    },
    tax_rate: {
      type: DataTypes.DECIMAL(5, 2),
    },
    tax_amount: {
      type: DataTypes.DECIMAL(15, 2),
    },
    discount_rate: {
      type: DataTypes.DECIMAL(5, 2),
    },
    discount_amount: {
      type: DataTypes.DECIMAL(15, 2),
    },
    received_quantity: {
      type: DataTypes.DECIMAL(15, 3),
      defaultValue: 0.000,
    },
    notes: {
      type: DataTypes.TEXT,
    }
  }, {
    tableName: 'po_items',
    timestamps: false
  });

  const SupplierRating = sequelize.define('SupplierRating', {
    rating_id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    supplier_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    po_id: {
      type: DataTypes.INTEGER,
    },
    quality_score: {
      type: DataTypes.DECIMAL(3, 2),
    },
    delivery_score: {
      type: DataTypes.DECIMAL(3, 2),
    },
    price_score: {
      type: DataTypes.DECIMAL(3, 2),
    },
    communication_score: {
      type: DataTypes.DECIMAL(3, 2),
    },
    overall_score: {
      type: DataTypes.DECIMAL(3, 2),
    },
    delivery_time_days: {
      type: DataTypes.INTEGER,
    },
    defect_rate: {
      type: DataTypes.DECIMAL(5, 2),
    },
    comments: {
      type: DataTypes.TEXT,
    },
    rated_by: {
      type: DataTypes.INTEGER,
    }
  }, {
    tableName: 'supplier_ratings',
    timestamps: true,
    createdAt: 'rated_at',
    updatedAt: false
  });

  return { Supplier, PurchaseOrder, POItem, SupplierRating };
};

File: backend/app/models/procurement.py
"""
Procurement Module Models
Suppliers, Purchase Orders, and related entities
"""

from sqlalchemy import Column, Integer, String, Numeric, Date, DateTime, ForeignKey, Text, Enum as SQLEnum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum
from app.core.database import Base

class SupplierStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    BLACKLISTED = "blacklisted"

class POStatus(str, enum.Enum):
    DRAFT = "draft"
    PENDING_APPROVAL = "pending_approval"
    APPROVED = "approved"
    ORDERED = "ordered"
    PARTIALLY_RECEIVED = "partially_received"
    RECEIVED = "received"
    CANCELLED = "cancelled"

class Supplier(Base):
    """Supplier master data"""
    __tablename__ = "suppliers"
    
    supplier_id = Column(Integer, primary_key=True, index=True)
    supplier_code = Column(String(50), unique=True, nullable=False, index=True)
    name = Column(String(200), nullable=False)
    contact_person = Column(String(100))
    email = Column(String(100))
    phone = Column(String(20))
    address = Column(Text)
    city = Column(String(100))
    state = Column(String(100))
    country = Column(String(100))
    postal_code = Column(String(20))
    tax_id = Column(String(50))
    payment_terms = Column(String(100))
    credit_limit = Column(Numeric(15, 2))
    rating = Column(Numeric(3, 2))  # 0.00 to 5.00
    status = Column(SQLEnum(SupplierStatus), default=SupplierStatus.ACTIVE)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    purchase_orders = relationship("PurchaseOrder", back_populates="supplier")
    supplier_ratings = relationship("SupplierRating", back_populates="supplier")

class PurchaseOrder(Base):
    """Purchase orders"""
    __tablename__ = "purchase_orders"
    
    po_id = Column(Integer, primary_key=True, index=True)
    po_number = Column(String(50), unique=True, nullable=False, index=True)
    supplier_id = Column(Integer, ForeignKey("suppliers.supplier_id"), nullable=False)
    order_date = Column(Date, nullable=False, index=True)
    expected_delivery_date = Column(Date)
    actual_delivery_date = Column(Date)
    status = Column(SQLEnum(POStatus), default=POStatus.DRAFT, index=True)
    total_amount = Column(Numeric(15, 2))
    tax_amount = Column(Numeric(15, 2))
    discount_amount = Column(Numeric(15, 2))
    net_amount = Column(Numeric(15, 2))
    payment_terms = Column(String(100))
    notes = Column(Text)
    approved_by = Column(Integer, ForeignKey("users.user_id"))
    approved_at = Column(DateTime(timezone=True))
    created_by = Column(Integer, ForeignKey("users.user_id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    supplier = relationship("Supplier", back_populates="purchase_orders")
    items = relationship("POItem", back_populates="purchase_order", cascade="all, delete-orphan")

class POItem(Base):
    """Purchase order line items"""
    __tablename__ = "po_items"
    
    item_id = Column(Integer, primary_key=True, index=True)
    po_id = Column(Integer, ForeignKey("purchase_orders.po_id"), nullable=False)
    material_id = Column(Integer, ForeignKey("materials.material_id"), nullable=False)
    quantity = Column(Numeric(15, 3), nullable=False)
    unit_price = Column(Numeric(15, 2), nullable=False)
    total_price = Column(Numeric(15, 2))
    tax_rate = Column(Numeric(5, 2))
    tax_amount = Column(Numeric(15, 2))
    discount_rate = Column(Numeric(5, 2))
    discount_amount = Column(Numeric(15, 2))
    received_quantity = Column(Numeric(15, 3), default=0)
    notes = Column(Text)
    
    # Relationships
    purchase_order = relationship("PurchaseOrder", back_populates="items")
    material = relationship("Material")

class SupplierRating(Base):
    """Supplier performance ratings"""
    __tablename__ = "supplier_ratings"
    
    rating_id = Column(Integer, primary_key=True, index=True)
    supplier_id = Column(Integer, ForeignKey("suppliers.supplier_id"), nullable=False)
    po_id = Column(Integer, ForeignKey("purchase_orders.po_id"))
    quality_score = Column(Numeric(3, 2))  # 0-5
    delivery_score = Column(Numeric(3, 2))  # 0-5
    price_score = Column(Numeric(3, 2))  # 0-5
    communication_score = Column(Numeric(3, 2))  # 0-5
    overall_score = Column(Numeric(3, 2))  # 0-5
    delivery_time_days = Column(Integer)
    defect_rate = Column(Numeric(5, 2))  # Percentage
    comments = Column(Text)
    rated_by = Column(Integer, ForeignKey("users.user_id"))
    rated_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    supplier = relationship("Supplier", back_populates="supplier_ratings")
